modules/tinymce/src/core/main/ts/dom/Empty.ts

```diff
@@ -1,85 +1,119 @@
+interface IsEmptyOptions {
+  readonly skipBogus?: boolean;
+  readonly includeZwsp?: boolean;
+  readonly checkRootAsContent?: boolean;
+  readonly isContent?: (node: Node) => boolean;
+}
+
+type IsContentOptions = Pick<IsEmptyOptions, 'includeZwsp' | 'isContent'>;
+
+const defaultOptionValues: IsEmptyOptions = {
+  skipBogus: true,
+  includeZwsp: false,
+  checkRootAsContent: false,
 };
 
-const isWhitespace = (node: Node, rootNode: Node): boolean => {
-  return NodeType.isText(node) && isWhitespaceText(node.data) && !hasWhitespacePreserveParent(node, rootNode);
+const hasWhitespacePreserveParent = (node: Node, rootNode: Node, schema: Schema): boolean => {
+  const rootElement = SugarElement.fromDom(rootNode);
+  const startNode = SugarElement.fromDom(node);
+  const whitespaceElements = schema.getWhitespaceElements();
+  const predicate = (node: SugarElement<Node>) => Obj.has(whitespaceElements, SugarNode.name(node));
+  return PredicateExists.ancestor(startNode, predicate, Fun.curry(Compare.eq, rootElement));
 };
 
 const isNamedAnchor = (node: Node): boolean => {
   return NodeType.isElement(node) && node.nodeName === 'A' && !node.hasAttribute('href') && (node.hasAttribute('name') || node.hasAttribute('id'));
 };
 
-const isContent = (node: Node, rootNode: Node): boolean => {
-  return (CaretCandidate.isCaretCandidate(node) && !isWhitespace(node, rootNode)) || isNamedAnchor(node) || isBookmark(node);
+const isNonEmptyElement = (node: Node, schema: Schema): boolean => {
+  return NodeType.isElement(node) && Obj.has(schema.getNonEmptyElements(), node.nodeName);
 };
 
 const isBookmark = NodeType.hasAttribute('data-mce-bookmark');
-const isBogus = NodeType.hasAttribute('data-mce-bogus');
-const isBogusAll = NodeType.hasAttributeValue('data-mce-bogus', 'all');
 
 const hasNonEditableParent = (node: Node) =>
   Traverse.parentElement(SugarElement.fromDom(node)).exists((parent) => !ContentEditable.isEditable(parent));
 
-const isEmptyNode = (targetNode: Node, skipBogus: boolean): boolean => {
+const isWhitespace = (node: Text, rootNode: Node, schema: Schema): boolean =>
+  isWhitespaceText(node.data)
+  && !hasWhitespacePreserveParent(node, rootNode, schema);
+
+const isText = (node: Node, rootNode: Node, schema: Schema, options: IsEmptyOptions) =>
+  NodeType.isText(node)
+  && !isWhitespace(node, rootNode, schema)
+  && (!options.includeZwsp || !isZwsp(node.data));
+
+const isContentNode = (schema: Schema, node: Node, rootNode: Node, options: IsContentOptions): boolean => {
+  return Type.isFunction(options.isContent) && options.isContent(node)
+  || isNonEmptyElement(node, schema)
+  || isBookmark(node)
+  || isNamedAnchor(node)
+  || isText(node, rootNode, schema, options)
+  || NodeType.isContentEditableFalse(node)
+  || NodeType.isContentEditableTrue(node) && hasNonEditableParent(node);
+};
+
+const isEmptyNode = (schema: Schema, targetNode: Node, options?: IsEmptyOptions): boolean => {
+  options = { ...defaultOptionValues, ...options };
```

Comment:
Please don't assign to function arguments. I'm surprised our linter allows this.

I think this ES6 trick is a nicer way to do it than double-spread
`const { skipBogus = true, includeZwsp = false, checkRootAsContent: false } = options`

Although you'd have to rebuild the `options` object later on, I suppose, but that still seems better.

Reply:
I keep the default values outside of function since it is used in several places. I will just use another name for the argument. IMO, `options.someKey` is easy to get where the thing comes from.